src/twiliosim.cr

```diff
@@ -17,99 +23,155 @@ class Twiliosim::Server
     when %r(/Accounts/(.+)/Calls.*)
       account_sid = $1
-      body = request.body
-      unless body
-        plain_response(context, 400, "Request body is missing")
-        return
-      end
-      body = body.gets_to_end
-      unless body
-        plain_response(context, 400, "Request body is missing")
-        return
-      end
-      if body.blank?
-        plain_response(context, 400, "Request body is missing")
-        return
-      end
-      params = HTTP::Params.parse(body)
-      unless params.has_key?("Url")
-        plain_response(context, 400, "Body param 'Url' is missing")
-        return
-      end
-      verboice_url = params["Url"]
-      unless params.has_key?("From")
-        plain_response(context, 400, "Body param 'From' is missing")
-        return
+      handle_call_request(context, account_sid)
+    else
+      plain_response(context, 404, "404 Not Found")
+    end
+  end
+
+  private def handle_incoming_phone_numbers_request(context)
+    context.response.status_code = 200
+    context.response.content_type = "application/json"
+    response = {"sid" => UUID.random().to_s()}
+    response.to_json(context.response)
+  end
+
+  private def handle_call_request(context : HTTP::Server::Context, account_sid : String)
+    unknown_error_message = "Internal error getting the request params"
+
+    begin
+      body_params = get_body_params(context.request.body, ["From", "To", "Url"])
+    rescue ex: BadRequestException
+      message = ex.message
+      unless message
+        puts "BadRequestException message is missing"
+        raise unknown_error_message
       end
-      from = params["From"]
-      unless params.has_key?("From")
-        plain_response(context, 400, "Body param 'To' is missing")
+        plain_response(context, 400, message)
+      return
+    end
+
+    from = body_params["From"]
+    to = body_params["To"]
+    verboice_url = body_params["Url"]
+
+    unless from && to && verboice_url
+      puts "Required (and validated) param is missing"
+      raise unknown_error_message
+    end
+
+    call = create_and_start_call(to, from, account_sid)
+    response_call_created(context, call.id)
+    spawn do
+      # We give Verboice a sec to process the response. Just in case it needs it.
+      sleep 1.seconds
+
+      # verboice_url cannot be nil, so the `not_nil!` call shouldn't be here.
+      # It's included here just because of [this Crystal compiler known issue](https://github.com/crystal-lang/crystal/issues/3093)
+      handle_created_call(verboice_url.not_nil!, call)
+    end
+  end
+
+  private def create_and_start_call(to : String, from : String, @account_sid : String) : TwilioCall
+    call = @db.create_call(to, from, account_sid)
+    call.start()
+    @db.update_call(call)
+  end
+
+  private def finish_and_update_call(call : TwilioCall) : TwilioCall
+    call.finish()
+    @db.update_call(call)
+  end
+
+  private def get_body_params(body : IO | Nil, required_params : Array) : HTTP::Params
```

Comment:
Better make `get_body_params` return a `{to: String,  from: String, url: String}`, that will remove some unneeded checks in `handle_call_request`